iree/turbine/kernel/wave/index_sequence_analysis.py

```diff
@@ -462,6 +379,226 @@ def set_node_index(
+            node.vector_shapes,
+        ),
+        (
+            node,
+            specialize_index(index, {MMA_LHS: 0, MMA_RHS: 0, MMA_ACC: 1}),
+            node.vector_shapes,
+        ),
+    ]
+
+
+def populate_non_mma_sources(
+    node: Read | Write,
+    hardware_constraint: HardwareConstraint,
+    workgroup_constraints: list[WorkgroupConstraint],
+):
+    """
+    Initialize the sources with the read and/or write nodes
+    and their index sequences and vector shapes. These will
+    be propagated to the rest of the graph.
+    """
+    index: dict[IndexSymbol, IndexSequence] = {}
+    for dim in node.indexing_dims:
+        elements_per_thread = (
+            1
+            if not is_contiguous_dim(
+                dim, node.indexing_dims, hardware_constraint.vector_shapes
+            )
+            else node.elements_per_thread
+        )
+        stride = compute_stride(
+            node.indexing_dims, hardware_constraint.vector_shapes, dim
+        )
+        wg_constraint = [x for x in workgroup_constraints if x.dim == dim]
+        if not wg_constraint:
+            continue
+        index[dim] = hardware_constraint.apply(
+            dim,
+            wg_constraint[0].workgroup_dim,
+            elements_per_thread,
+            stride,
+            False,
+            None,
+        )
+    return [(node, index, hardware_constraint.vector_shapes)]
+
+
+def combine_indices(
+    thread_independent_index: dict[IndexSymbol, IndexSequence],
+    thread_dependent_index: dict[IndexSymbol, IndexSequence],
+) -> dict[IndexSymbol, IndexSequence]:
+    combined_index = {k: v for k, v in thread_independent_index.items()}
+    for k in combined_index:
+        if k in thread_dependent_index:
+            combined_index[k].start += thread_dependent_index[k].start
+            combined_index[k].size = thread_dependent_index[k].size
+            combined_index[k].stride = thread_dependent_index[k].stride
+    return combined_index
+
+
+def add_nodes_to_sources(
+    source: CustomOp,
+    reduction: Reduction,
+    fn: Callable,
+    source_index: dict[IndexSymbol, IndexSequence],
+    source_vector_shapes: dict[IndexSymbol, int],
+    sources: list[
+        tuple[CustomOp, dict[IndexSymbol, IndexSequence], dict[IndexSymbol, int]]
+    ],
+) -> tuple[list[CustomOp], Reduction]:
+    """
+    Populate the sources with the inputs and users of the source node.
+    """
+    for args, reduction in [fn(source.fx_node, reduction)]:
+        logger.debug(f"{source.fx_node} -> {args}")
+        if not args:
+            break
+        for arg in args:
+            custom = get_custom(arg)
+            if isinstance(custom, (Allocate, Placeholder)) and not isinstance(
+                custom, IterArg
+            ):
+                continue
+            vector_shapes = (
+                custom.vector_shapes if custom.vector_shapes else source_vector_shapes
+            )
+            sources.append((custom, source_index, vector_shapes))
```

Comment:
why not use source.index? Seems like the only time sourc.index != source_index is when source.index has underlying indices before getting propagated. But wouldn't we want that?

Reply:
So source.index contains the unified index which is the sum of the thread dependent index and the thread independent index. During propagation we only want to propagate the thread dependent index and that's why we only propagate source_index.